vividus-extension-selenium/src/main/java/org/vividus/monitor/AbstractPublishingAttachmentOnFailureMonitor.java

```diff
@@ -0,0 +1,135 @@
+{
+    private static final Logger LOGGER = LoggerFactory.getLogger(AbstractPublishingAttachmentOnFailureMonitor.class);
+
+    private final ThreadLocal<Boolean> publishAttachmentOnFailureEnabled = ThreadLocal.withInitial(() -> Boolean.FALSE);
+
+    private final RunContext runContext;
+    private final IWebDriverProvider webDriverProvider;
+    private final EventBus eventBus;
+    private final String metaNameToSkipPublishing;
+
+    public AbstractPublishingAttachmentOnFailureMonitor(RunContext runContext, IWebDriverProvider webDriverProvider,
+            EventBus eventBus, String metaNameToSkipPublishing)
+    {
+        this.runContext = runContext;
+        this.webDriverProvider = webDriverProvider;
+        this.eventBus = eventBus;
+        this.metaNameToSkipPublishing = metaNameToSkipPublishing;
+    }
+
+    @Override
+    public void beforePerforming(String step, boolean dryRun, Method method)
+    {
+        if (isPublishingEnabled(method) && !isPublishingDisabledInStoryMeta()
+                && !isPublishingDisabledInScenarioMeta())
+        {
+            publishAttachmentOnFailureEnabled.set(Boolean.TRUE);
+        }
+    }
+
+    @Override
+    public void afterPerforming(String step, boolean dryRun, Method method)
+    {
+        if (isPublishingEnabled(method))
+        {
+            publishAttachmentOnFailureEnabled.set(Boolean.FALSE);
+        }
+    }
+
+    @Subscribe
+    public void onAssertionFailure(AssertionFailedEvent event)
+    {
+        if (publishAttachmentOnFailureEnabled.get() && webDriverProvider.isWebDriverInitialized())
+        {
+            publishAttachment();
+        }
+    }
+
+    private boolean isPublishingDisabledInStoryMeta()
+    {
+        RunningStory runningStory = runContext.getRunningStory();
+        return runningStory.getStory().getMeta().hasProperty(metaNameToSkipPublishing);
+    }
+
+    private boolean isPublishingDisabledInScenarioMeta()
+    {
+        return Optional.of(runContext.getRunningStory()).map(RunningStory::getRunningScenario)
+                .map(RunningScenario::getScenario).map(Scenario::getMeta)
+                .map(m -> m.hasProperty(metaNameToSkipPublishing)).orElse(Boolean.FALSE);
+    }
+
+    protected abstract boolean isPublishingEnabled(Method method);
+
+    protected abstract void publishAttachment();
+
+    protected void publishAttachment(byte[] content, String fileName)
+    {
+        Attachment attachment = new Attachment(content, fileName);
+        eventBus.post(new AttachmentPublishEvent(attachment));
+    }
+
+    protected Annotation getAnnotation(Method method, Class<? extends Annotation> annotationClass)
+    {
+        if (method != null)
+        {
+            AnnotatedElement annotatedElement = method.isAnnotationPresent(annotationClass) ? method
+                    : method.getDeclaringClass();
+            return annotatedElement.getAnnotation(annotationClass);
+        }
+        return null;
```

Comment:
does it make sense to operate here using Optionals not nulls?

Reply:
out of scope for this PR